indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java

```
@@ -0,0 +1,270 @@
+                );
+              }
+            }
+          }
+        },
+        intermediaryPartitionDiscoveryPeriodSec,
+        intermediaryPartitionDiscoveryPeriodSec,
+        TimeUnit.SECONDS
+    );
+    // Check supervisorTask status if its partitions have not been accessed in timeout.
+    // Delete partitions if the supervisorTask is already finished.
+    // Note that the overlord sends a cleanup request when a supervisorTask is finished. The below check is to trigger
+    // the self-cleanup for when middleManager misses the cleanup request from the overlord.
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
+          final DateTime now = DateTimes.nowUtc();
+          final Set<String> expiredSupervisorTasks = new HashSet<>();
+          for (Entry<String, DateTime> entry : supervisorTaskCheckTimes.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final DateTime checkTime = entry.getValue();
+            if (checkTime.isAfter(now)) {
+              expiredSupervisorTasks.add(supervisorTaskId);
+            }
+          }
+
+          final Map<String, TaskStatus> taskStatuses = indexingServiceClient.getTaskStatuses(expiredSupervisorTasks);
+          RuntimeException exception = null;
+          for (Entry<String, TaskStatus> entry : taskStatuses.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final TaskStatus status = entry.getValue();
+            if (status.getStatusCode().isComplete()) {
+              try {
+                deletePartitions(supervisorTaskId);
+              }
+              catch (IOException e) {
+                if (exception == null) {
+                  exception = new RuntimeException(e);
+                } else {
+                  exception.addSuppressed(e);
+                }
+              }
+            }
+          }
+
+          if (exception != null) {
+            throw exception;
```

this code is running in a scheduler , why do we throw the overall exception in the end instead of logging them as they happened ?

---

can't we log the exceptions as they occur instead of accumulating all of them inside one and then logging it in the end ?

---

Ah sure, fixed.